enhancements/scheduling/load-aware-vm-rebalancing-through-descheduler.md

```diff
@@ -0,0 +1,472 @@
+* **Limited eviction abilities**: Not all resource utilization metrics can determine which pod is causing resource pressure or predict the impact of an eviction. As a result, only the lowest-priority pod is evicted during each descheduling cycle, relying on the cluster to adapt to the change and reflect it back through the metrics. To ensure a quick response to cluster changes, the next descheduling cycle should run within minutes. As a consequence, strategies for evicting multiple pods while achieving more efficient rebalancing are beyond the current scope. With one exception of blindly evicting multiple pods when tuned.
+* **Control-plane nodes are excluded from VM scheduling**: VMs are intended to be scheduled only to worker nodes. While an OpenShift cluster can enable scheduling to control-plane nodes, this functionality is not recommended, as it could unexpectedly affect the scheduling of control plane pods due to soft-tainting.
+* **Distributive nature of eviction, scheduling and node soft-tainting**: Eviction, scheduling, and node tainting are independent processes. When a VM pod is evicted and a live migration is initiated, its replacement pod is scheduled to a new node based on past node utilization (as the evicted pod still exists in the system). Similarly, node soft-tainting responds to the cluster state recorded before the VM pod eviction was started. This inherent nature of the process cannot be fully resolved but can be regulated by fine-tuning various parameters.
+* **Hard-tainting**: hard taints (`effect: NoSchedule`/`effect: NoExecute`)  can restrict scheduling decisions but may also reduce overall cluster flexibility. The node controller already applies certain condition-based hard taints for critical node states such as `MemoryPressure`, `DiskPressure`, and `PIDPressure`. However, these hard taints are only set when a node reaches a serious or critical condition. The goal of this proposal is to adopt a similar approach using a soft taint when a node is overutilized but not yet in a critical state. This ensures that the scheduler is aware of resource pressure without making the node entirely unschedulable.
+
+## Proposal
+
+The enhancements will include the following functionality changes:
+1. Updating the descheduler to be load-aware to evict pods based on actual resource utilization.
+2. Soft-tainting over-utilized nodes so the requested resources are less effective during scheduling.
+3. Extending the descheduler operator customization to enable the load-aware eviction with tuning options.
+
+### Update LowNodeUtilization plugin to consume Prometheus metrics
+
+Currently, the descheduler's `LowNodeUtilization` plugin classifies node utilization based on Kubernetes metrics. However, these metrics primarily capture CPU and memory usage at the node and pod levels, which are not well-suited for VMs. By consuming Prometheus metrics, the plugin can leverage a broader range of data that better aligns with VM-specific needs.
+
```

Comment:
switching from metrics API to consuming raw Prometheus metrics, how do we ensure that the right version of the components that provide these metrics are available? if not available, then do we have a plan to surface this to the admin in any status object?

Also, for future, is it feasible to abstract these metrics into the metrics API?